In [1]:
import pandas as pd

# Cleaning variables

Users of an online entry system used to have the ability to freely type in responses to questions. This is causing issues when trying to analyze the adoptable dogs dataset, dogs. Here is the current frequency table of the "sex" column:

male      1672

female    1249

 MALE        10
 
 FEMALE       5
 
Malez  

   

* Now that the system only takes responses of "female" and "male", you want this variable to match the updated
* Update the misspelled response "Malez" to be "male" by creating the replacement map, ment map, replace_map.


In [2]:
dogs = pd.read_csv("/kaggle/input/adoptable-dogs/ShelterDogs.csv")
dogs.head()

ID            name   age     sex               breed  date_found  \
0  23807            Gida  0.25  female         Unknown Mix  2019-12-10   
1    533  Frida És Ricsi  0.17  female         Unknown Mix  2019-12-01   
2  23793             NaN  4.00    male         Unknown Mix  2019-12-08   
3  23795             NaN  1.00    male         Unknown Mix  2019-12-08   
4  23806             Amy  2.00  female  French Bulldog Mix  2019-12-10   

  adoptable_from      posted            color    coat    size neutered  \
0     2019-12-11  2019-12-11              red   short   small       no   
1     2019-12-01  2019-12-09  black and white   short   small       no   
2     2019-12-23  2019-12-08      saddle back   short  medium       no   
3     2019-12-23  2019-12-08     yellow-brown  medium  medium       no   
4     2019-12-11  2019-12-11            black   short   small       no   

  housebroken likes_people likes_children get_along_males get_along_females  \
0         NaN          NaN            NaN             NaN               NaN   
1         NaN          yes            yes             yes               yes   
2         NaN          NaN            NaN             NaN               NaN   
3         NaN          NaN            NaN             NaN               NaN   
4         NaN          NaN            NaN             NaN               NaN   

  get_along_cats keep_in  
0            NaN     NaN  
1            yes     NaN  
2            NaN     NaN  
3            NaN     NaN  
4            NaN     NaN

In [3]:
# Fix the misspelled word 
replace_map =  {"Malez":"Male"}
dogs["sex"].replace(replace_map, inplace = True)

/tmp/ipykernel_30/3546149000.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dogs["sex"].replace(replace_map, inplace = True)


In [4]:
# Update the misspelled response "Malez" to be "male" by creating the replacement map,
# replace_map.

# Fix the misspelled word
replace_map = {"Malez": "male"}

# Update the sex column using the created map
dogs["sex"] = dogs["sex"].replace(replace_map)

# Strip away leading whitespace
dogs["sex"] = dogs["sex"].str.strip()

# Make all responses lowercase
dogs["sex"] = dogs["sex"].str.lower()

# Convert to a categorical Series
dogs["sex"] = dogs["sex"].astype("category")

print(dogs["sex"].value_counts())

sex
male      1681
female    1256
Name: count, dtype: int64


 Categorical variables are usually just strings with some additional properties. An easy way to update them is using .str. Just don't forget to convert the column back to a categorical Series!

# Accessing and filtering data

You are working on a Python application to display information about the dogs available for adoption at your local animal shelter. Some of the variables of interest, such as "breed", "size", and "coat", are saved as categorical variables. In order for this application to work properly, you need to be able to access and filter data using these columns.

The ID variable has been set as the index of the pandas DataFrame dogs.



* Print the "coat" value for the dog with an ID of 23807.
* For dogs with a long "coat", print the number of each "sex".
* Print the average age of dogs with a "breed" of "English Cocker Spaniel".
* Filter to the dogs with "English" in their "breed" name using the .contains() method.


In [5]:
# Print the category of the coat for ID 118
print(dogs.loc[118, "coat"])

short


In [6]:
# Find the count of male and female dogs who have a "long" coat
print(dogs.loc[dogs["coat"] == "long", "sex"].value_counts())

sex
male      124
female     56
Name: count, dtype: int64


In [7]:
# Print the mean age of dogs with a breed of "English Cocker Spaniel"
print(dogs.loc[dogs["breed"] == "English Cocker Spaniel", "age"].mean())

8.186153846153847


In [8]:
# Count the number of dogs that have "English" in their breed name
print(dogs[dogs["breed"].str.contains("English", regex=False)].shape[0])

35
